src/VisualStudio/Xaml/Impl/Implementation/LanguageServer/XamlLanguageServerProtocol.cs

```diff
@@ -19,11 +22,26 @@ namespace Microsoft.VisualStudio.LanguageServices.Xaml.LanguageServer
     [Export(typeof(XamlLanguageServerProtocol))]
     internal sealed class XamlLanguageServerProtocol : AbstractRequestHandlerProvider
     {
+        private readonly XamlProjectService _projectService;
+
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public XamlLanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers)
+        public XamlLanguageServerProtocol([ImportMany] IEnumerable<Lazy<IRequestHandler, IRequestHandlerMetadata>> requestHandlers, XamlProjectService projectService)
             : base(requestHandlers, languageName: StringConstants.XamlLanguageName)
         {
+            _projectService = projectService;
+        }
+
+        protected override Task<ResponseType> ExecuteRequestAsync<RequestType, ResponseType>(RequestExecutionQueue queue, RequestType request, ClientCapabilities clientCapabilities, string? clientName, string methodName, bool mutatesSolutionState, IRequestHandler<RequestType, ResponseType> handler, CancellationToken cancellationToken)
+        {
+            var textDocument = handler.GetTextDocumentIdentifier(request);
+
+            if (textDocument is { Uri: { IsAbsoluteUri: true } documentUri })
```

Comment:
@mgoertz-msft would notifying the project service to track a document be done better in a didOpen LSP handler?  That way we don't have to expose this ExecuteRequestAsync method as virtual.

Though we would still want to have our didOpen handler run as well to ensure we're using the consistent LSP snapshot state.

Reply:
I suppose that would work. Does your didOpen handler even get called for XAML files right now? XAML doesn't actually have one.

If XAML needs one for this and needs to call some code in yours then that code would need to be factored out first so it can be called by both of us.

Reply:
It looks like no - http://sourceroslyn.io/#Microsoft.VisualStudio.LanguageServices.Xaml/Implementation/LanguageClient/XamlInProcLanguageClient.cs,55

I originally thought we did (and therefore we were also creating a xaml LSP solution snapshot and sync'ing separately) which is why I mentioned needing to call our own handler.  So since we aren't handling text doc sync for you it may be as easy as setting OpenClose to true and then creating your own handler implementation.  I don't think there's any rush on doing that though.

I don't know how easy it is for xaml, but at some point you may want to consider having a separate xaml LSP snapshot similar to what we do to avoid synchronization issues - the text buffer changes are not necessarily 100% ordered in relation to LSP messages, so you can encounter various race conditions if you aren't using LSP text sync.

Reply:
I must admit that LSP text syncing is still a bit of a mystery to me. I keep hearing about potential synchronization issues, but I have yet to see any concrete scenarios where that occurs. Of course, it could just be that we haven't run into any of that because we are still working on enabling/refining the most basic language service scenarios.